File: backend/src/com/origin/AccountCache.java
package com.origin;

import com.origin.entity.Account;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.ReentrantLock;

/**
 * локальный кэш залогиненных юзеров
 */
public class AccountCache
{
	private static final int TIMEOUT_LOCK = 1000;

	/**
	 * храним объекты юзеров в памяти
	 */
	private final ConcurrentHashMap<String, Account> _accounts = new ConcurrentHashMap<>();

	/**
	 * блокировка на любые операции с кэшем
	 */
	private final ReentrantLock _lock = new ReentrantLock();

	public void drop(String ssid)
	{
		_accounts.remove(ssid);
	}

	public String drop(Account account)
	{
		for (Map.Entry<String, Account> entry : _accounts.entrySet())
		{
			if (entry.getValue().getId() == account.getId())
			{
				String ssid = entry.getKey();
				_accounts.remove(ssid);
				return ssid;
			}
		}
		return null;
	}

	/**
	 * добавить юзера в кэш с одновременной его авторизацией
	 */
	public boolean addWithAuth(Account account) throws InterruptedException
	{
		// операции добавления юзера в кэш позволим делать только по 1 одновременно
		// поэтому используем блокировку
		if (_lock.tryLock(TIMEOUT_LOCK, TimeUnit.MILLISECONDS))
		{
			try
			{
				if (_accounts.containsValue(account))
				{
					return false;
				}
				account.generateSessionId();
				_accounts.put(account.getSsid(), account);
				return true;
			}
			finally
			{
				_lock.unlock();
			}
		}
		return false;
	}

	public void addUser(Account account) throws InterruptedException
	{
		if (_lock.tryLock(TIMEOUT_LOCK, TimeUnit.MILLISECONDS))
		{
			// TODO
		}
	}
}

File: backend/src/com/origin/AccountCache.kt
package com.origin

import com.origin.jooq.tables.records.AccountRecord
import com.origin.util.LockByName
import java.util.concurrent.ConcurrentHashMap

/**
 * локальный кэш залогиненных юзеров
 * 1. у нас ровно один гейм севрер сервер. тогда храним тут мапу соответствия ssid->account entity
 * в кэше всегда актуальная сохраненная сущность. гарантируем что один и тот же юзер может авторизоваться только 1 раз
 * 2. у нас кластер и гейм серверов несколько. тогда хранить сущность в кэше нельзя, т.к. авторизоваться
 * аккаунт может на другом сервере и у нас будут не актуальные данные. в этом случае надо при каждом запросе грузить
 * аккаунт из базы (сверять ssid из запроса с ssid из базы), а локально вообще ничего не хранить. смысла в кэше в этом случае нет.
 */
class AccountCache {
    /**
     * храним объекты юзеров в памяти
     * ssid > Account
     */
    private val accountBySsid = ConcurrentHashMap<String, AccountRecord>()

    /**
     * храним сессии по ид юзера
     */
    private val accountSsidById = HashMap<Long, String>()

    /**
     * блокировки по ид юзера
     */
    private val namedLocks = LockByName<Long>()

    /**
     * добавить юзера в кэш с одновременной его авторизацией
     *
     * @return true если успешно добавили и в кэше НЕ было такого ssid, false в случае если коллизия ssid
     */
    fun add(account: AccountRecord) {
        val accountId = account.id ?: throw RuntimeException("no account id")
        // РЕАЛИЗАЦИЯ варианта 1 (описание выше)
        // заблокируем по ид аккаунта, дабы нельзя было одному акканту одновременно делать НЕ атомарные операции ниже
        val lock = namedLocks.getLock(accountId)
        try {
            lock.lock()
            // ищем старую сессию для этого аккаунта
            val oldSsid = accountSsidById[account.id]
            // если сессия найдена
            if (oldSsid != null) {
                // удалим старую сессию
                accountBySsid.remove(oldSsid)
            }
            // запомним новую сессию
            accountBySsid[account.ssid!!] = account
            accountSsidById[accountId] = account.ssid!!
        } finally {
            lock.unlock()
        }
    }

    fun get(ssid: String?): AccountRecord? {
        // РЕАЛИЗАЦИЯ варианта 1 (описание выше)
        return if (ssid == null) null else accountBySsid[ssid]
    }
}
